ostd/src/arch/riscv/mm/mod.rs

```diff
@@ -0,0 +1,241 @@
+        const PBMT_IO =         1 << 62;
+        /// Naturally aligned power-of-2
+        const NAPOT =           1 << 63;
+    }
+}
+
+pub(crate) fn tlb_flush_addr(vaddr: Vaddr) {
+    unsafe {
+        riscv::asm::sfence_vma(0, vaddr);
+    }
+}
+
+pub(crate) fn tlb_flush_addr_range(range: &Range<Vaddr>) {
+    for vaddr in range.clone().step_by(PAGE_SIZE) {
+        tlb_flush_addr(vaddr);
+    }
+}
+
+pub(crate) fn tlb_flush_all_excluding_global() {
+    // TODO: excluding global?
+    riscv::asm::sfence_vma_all()
+}
+
+pub(crate) fn tlb_flush_all_including_global() {
+    riscv::asm::sfence_vma_all()
+}
+
+#[derive(Clone, Copy, Pod, Default)]
+#[repr(C)]
+pub struct PageTableEntry(usize);
+
+/// Activate the given level 4 page table.
+///
+/// "satp" register doesn't have a field that encodes the cache policy,
+/// so `_root_pt_cache` is ignored.
+///
+/// ## Safety
+///
+/// Changing the level 4 page table is unsafe, because it's possible to violate memory safety by
+/// changing the page mapping.
+pub unsafe fn activate_page_table(root_paddr: Paddr, _root_pt_cache: CachePolicy) {
+    assert!(root_paddr % PagingConsts::BASE_PAGE_SIZE == 0);
+    let ppn = root_paddr >> 12;
+    riscv::register::satp::set(riscv::register::satp::Mode::Sv48, 0, ppn);
+}
+
+pub fn current_page_table_paddr() -> Paddr {
+    riscv::register::satp::read().ppn() << 12
+}
+
+impl PageTableEntry {
+    const PHYS_ADDR_MASK: usize = 0x003F_FFFF_FFFF_FC00;
+
+    fn new_paddr(paddr: Paddr) -> Self {
+        let ppn = paddr >> 12;
+        Self(ppn << 10)
+    }
+}
+
+impl PageTableEntryTrait for PageTableEntry {
+    fn is_present(&self) -> bool {
+        self.0 & PageTableFlags::VALID.bits() != 0
+    }
+
+    fn new_page(paddr: Paddr, _level: PagingLevel, prop: PageProperty) -> Self {
+        let mut pte = Self::new_paddr(paddr);
+        pte.set_prop(prop);
+        pte
+    }
+
+    fn new_pt(paddr: Paddr) -> Self {
+        // In RISC-V, non-leaf PTE should have RWX = 000,
+        // and D, A, and U are reserved for future standard use.
+        let pte = Self::new_paddr(paddr);
+        PageTableEntry(pte.0 | PageTableFlags::VALID.bits())
+    }
+
+    fn paddr(&self) -> Paddr {
+        let ppn = (self.0 & Self::PHYS_ADDR_MASK) >> 10;
+        ppn << 12
+    }
+
+    fn prop(&self) -> PageProperty {
+        let mut flags = PageFlags::empty();
+        let mut priv_flags = PrivFlags::empty();
+        if self.0 & PageTableFlags::READABLE.bits() != 0 {
+            flags |= PageFlags::R;
+        }
+        if self.0 & PageTableFlags::WRITABLE.bits() != 0 {
+            flags |= PageFlags::W;
+        }
+        if self.0 & PageTableFlags::EXECUTABLE.bits() != 0 {
+            flags |= PageFlags::X;
+        }
+        if self.0 & PageTableFlags::ACCESSED.bits() != 0 {
+            flags |= PageFlags::ACCESSED;
+        }
+        if self.0 & PageTableFlags::DIRTY.bits() != 0 {
+            flags |= PageFlags::DIRTY;
+        }
```

Comment:
The implementation in x86 have introduced a macro to make life easier:

https://github.com/asterinas/asterinas/blob/0a36760f7a42b7fe27735e29f971423f6ec4b1ca/ostd/src/arch/x86/mm/mod.rs#L150-L155

https://github.com/asterinas/asterinas/blob/0a36760f7a42b7fe27735e29f971423f6ec4b1ca/ostd/src/arch/x86/mm/mod.rs#L187-L191